test/parallel_api/numeric/numeric.ops/exclusive_scan_by_segment.pass.cpp

```diff
@@ -20,146 +20,194 @@
 #include "support/test_config.h"
 #include "support/utils.h"
 
-#include <iostream>
-#include <iomanip>
-
-template<typename _T1, typename _T2>
-void ASSERT_EQUAL(_T1&& X, _T2&& Y) {
-    if(X!=Y)
-        std::cout << "CHECK CORRECTNESS (EXCLUSIVE_SCAN_BY_SEGMENT): fail (" << X << "," << Y << ")" << std::endl;
-}
-
 #if TEST_DPCPP_BACKEND_PRESENT
 #include <CL/sycl.hpp>
 
-void test_with_buffers()
-{
-    // create a buffer, being responsible for moving data around and counting dependencies
-    sycl::buffer<uint64_t, 1> _key_buf{ sycl::range<1>(10) };
-    sycl::buffer<uint64_t, 1> _val_buf{ sycl::range<1>(10) };
-    sycl::buffer<uint64_t, 1> _res_buf{ sycl::range<1>(10) };
+using namespace oneapi::dpl::execution;
+#endif
+using namespace TestUtils;
 
+struct test_exclusive_scan_by_segment
+{
+    template <typename Accessor1, typename Accessor2, typename Accessor3, typename Size>
+    void
+    initialize_data(Accessor1 host_keys, Accessor2 host_vals, Accessor3 host_val_res, Size n)
```

Comment:
Please, consider https://github.com/oneapi-src/oneDPL/pull/247#discussion_r694930284 for *by_segment algorithms too. In that case we can perform segmented sorting of a generated sequence.

Reply:
I've added the TODO comment in this case as well.  My concern is that the current test exercises the case where we have multiple segments with the same key, the results of which must not be combined. Generating and sorting the sequence would result with a single segment per key value.

Reply:
We can sort parts of a sequence, e.g. sort(keys_begin, keys_middle), and then sort(keys_middle, keys_end). Of course, the logic may be more complex.